src/Api/Upload.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+namespace Kirby\Api;
+
+use Kirby\Cms\App;
+use Kirby\Filesystem\Dir;
+use Kirby\Filesystem\F;
+use Kirby\Toolkit\Str;
+
+/**
+ * The Upload class supports file uploads in the
+ * context of the API
+ *
+ * @package   Kirby Api
+ * @author    Nico Hoffmann <nico@getkirby.com>
+ * @link      https://getkirby.com
+ * @copyright Bastian Allgeier
+ * @license   https://getkirby.com/license
+ * @since     4.3.0
+ */
+class Upload
+{
+	/**
+	 * Handle chunked uploads by merging all chunks
+	 * in the tmp directory and only returning the new
+	 * $source path to the tmp file once complete
+	 */
+	public static function chunk(
+		Api $api,
+		string $source,
+		string $filename
+	): string|null {
+		// if the file is uploaded in chunks
+		if ($isChunk = $api->requestHeaders('X-Upload-Chunk')) {
+			// ensure the directory exists
+			Dir::make($dir = static::tmp());
+
+			// append chunk to tmp file with unique filename (id suffix)
+			$id     = $api->requestHeaders('X-Unique-Upload-Id');
+			$chunk  = fopen($source, 'r');
+			$source = $dir . '/' . $filename;
+			$tmp    = $source . '.' . $id;
+			$part   = fopen($tmp, 'a');
+			stream_copy_to_stream($chunk, $part);
+
+			// if not the last chunk, return early
+			if ($isChunk !== 'last') {
+				return null;
+			}
+
+			// remove id from tmp filename now the file is complete
+			rename($tmp, $source);
+		}
+
+		return $source;
+	}
+
+	/**
+	 * Clean up tmp directory of stale files
+	 */
+	public static function clear(): void
+	{
+		foreach (Dir::files($dir = static::tmp(), [], true) as $file) {
+			// remove any file that hasn't been altered
+			// in th last 24 hours
+			if (F::modified($file) < time() - 86400) {
+				F::remove($file);
+			}
+		}
+
+		if (Dir::isEmpty($dir) === true) {
+			Dir::remove($dir);
+		}
+	}
+
+	/**
+	 * Returns the ideal size for a file chunk
+	 */
+	public static function chunkSize(): int
+	{
+		$max = [
+			Str::toBytes(ini_get('upload_max_filesize')),
+			Str::toBytes(ini_get('post_max_size'))
+		];
+
+		// if server is behind a cloudflare proxy
+		if (isset($_SERVER['HTTP_CF_CONNECTING_IP']) === true) {
+			$max[] = Str::toBytes('100M');
+		}
+
+		// to be sure, only use 95% of the max possible upload size
+		return (int)floor(min($max) * 0.95);
+	}
+
+	/**
+	 * Returns root of directory used for
+	 * temporarily storing (incomplete) uploads
+	 */
+	public static function tmp(): string
+	{
+		return App::instance()->root('media') . '/uploads';
```

Comment:
Is media/uploads an issue as tmp directory as it could be accessible?
Is it especially problematic, that e.g. half a PHP/.exe file could be uploaded to the tmp directory? Rules only kick in after last chunk has been uploaded.

Reply:
I think it could be risky, but @lukasbestle should probably have a look at it.

Reply:
Alternatively, do you have ideas for alternatives. The sys tmp folders (as initially suggested on Notion) don't work as they don't persist across the multiple requests.

Reply:
Maybe we could upload to an invisible folder in `content` like `/content/.uploads`. That folder is already blocked in web access by our suggested default server configs. Also if someone has set up the content folder on a different filesystem, we'd avoid moving the file between different filesystems after the last chunk.

Reply:
In addition, it would be great to run at least basic security checks if not all file rules. Having a PHP or Phar file on the server would be game over. One can easily upload a full file and pretend that another chunk follows afterwards, but never send another chunk.

Reply:
I would like to away from the `content` folder. Cause our setup is that basically everything in `content` means something. We'd have to add in a few places code to ignore that directory etc. And I don't like to spread the feature code even more around the system. Could we rather use `site/cache` or so? `site` subfolders are also protected.

Not sure how to go about the file rules yet as they depend a lot on the filename, extension and already creating a File object.

Reply:
What about `site/temp`? It could also be useful for other things in the future and also devs could easily customize its path by overriding the root.

File rules: You are right, probably too complex. But then we should at least validate the full temp name (filename + unique ID) against our security blocklists.

Reply:
I know that quite some people started setting up a custom storage folder for all those "secret" writable files: 

- sessions
- cache
- accounts

I agree that the tmp folder should be in the site folder as well to keep it save. But then you would need to take care of file permissions in yet another spot. If we had such a single, writable storage folder inside of site, it would be a lot easier. Otherwise we would also need another check in the installer and the system view for the new tmp folder. 

But I guess moving to a unified storage folder is making this PR even more complex than it already is.

Reply:
@bastianallgeier that really sounds like it's own project

Reply:
Or we just use a subdirectory in `site/cache` for the tmp chunk uploads.

Reply:
Oh, I quite like the idea.

Reply:
Using the cache directory as an intermediate solution makes sense. Maybe we could then implement a new `temp` root separately for v5. That would then contain `cache`, `sessions` and `uploads`. Wouldn't put `accounts` there though as they are not ephemeral unlike the others.